core/src/main/java/io/grpc/ServerImpl.java

```diff
@@ -502,20 +508,27 @@ public ServerStreamListenerImpl(ServerCall.Listener<ReqT> listener, Future<?> ti
       }
 
       @Override
-      public void messageRead(final InputStream message) {
-        try {
-          if (cancelled) {
-            return;
-          }
-
-          listener.onPayload(method.parseRequest(message));
-        } finally {
+      public void messagesRead(final Iterator<InputStream> messages) {
+        RuntimeException exception = null;
+        while (messages.hasNext()) {
+          InputStream message = messages.next();
           try {
-            message.close();
-          } catch (IOException e) {
-            throw new RuntimeException(e);
+            if (!cancelled) {
+              listener.onPayload(method.parseRequest(message));
+            }
+          } finally {
+            try {
+              message.close();
+            } catch (IOException e) {
```

Comment:
I suspect we do this a lot ... maybe we should just make a utility method for closing input streams?